### PR TITLE
unix/main: Use standard pyexec/repl for unix port.

### PR DESCRIPTION
**Pull request by andrewleech**
This improves repl usage consistency across ports. 
Only enabled when MICROPY_USE_READLINE == 1 (default).

In particular I noticed that `aiorepl` didn't work properly on the unix port currently, it would echo the "just entered line" on a newline each time you hit enter, eg.
```
>>> import aiorepl, asyncio;asyncio.run(aiorepl.task())
Starting asyncio REPL...
--> print("hello")
print("hello")
hello
```
I feel there were also a few other inconsistencies with how I'm used to the repl working on the stm port / via mpremote which are alsos hopefully resolved now.

I've got an optional couple of lines to restore/reset stdio_raw mode when running code - however if raw mode is disabled the aiorepl issue above remains. I think for consistency with other ports it's better to keep stdio in raw mode, but not sure if this is likely to have other side effects?

This PR consilidates the repl functionality and fixes the above aiorepl glitch.

### CODE REVIEW

**Comment by dpgeorge** [comment]
Thanks for this, it's a good clean up.

I actually attempted it before, see #6008.  The approach here is much simpler, so maybe it's a good first step to make.

**Comment by andrewleech** [comment]
Ah interesting, I did see some of the return handling and kind of assumed it shouldn't matter - if someone is dealing with the repl (live coding) they're unlikely to care about the exit code?

I didn't consider the SystemExit interrupt at all, haven't looked at how that behaves but probably should!

If running a python file directly then certainly the exit code is important but that shouldn't be impacted by this.

**Comment by andrewleech** [comment]
Ah, I might need to read through the other PR discussion a bit more... but in the mean time this is a little off:
```
$ ~/micropython/ports/unix/build-standard/micropython
MicroPython v1.22.0-preview.32.g2c1c7d8b30 on 2023-10-26; linux [GCC 10.2.1] version
Type "help()" for more information.
>>> raise ValueError()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError:
>>> raise SystemExit()
>>>
```

**Comment by jimmo** [comment]
See also https://github.com/micropython/micropython/pull/12807

**Comment by andrewleech** [comment]
I ran into a tricky issue when trying to fix the unit tests; the banner on Unix port has changed a little with this MR (now consistent with other ports) so unit test exp needed to be updated to match.

I initially tried just copying the results/*.out files to replace the original .exp ones, but the tests still failed! There were inconsistent newlines in the files that tripped up the test.
I narrowed the program down to the pty usage:
* micropython was outputting stdout with '\r\n`
* pty was converting `\n` to `\r\n`, creating a few `\r\r\n`
* the run_tests script does convert `\r\n` back to `\n` but I think it was only doing that once before writing failed test *.out files, then converting a second time to do the actual comparison to *.exp. 
Or it was the offer way around... or maybe subprocess was doing some conversions too.
Either way I was getting some `\r\n` and some `\n` in the test out files.
The same issue is discussed here which is where I found reference to how to disable the pty below conversions: https://www.thecodingforums.com/threads/new-line-conversion-with-popen-attached-to-a-pty.961960/post-5100925

File: ports/windows/micropython.vcxproj [hunk 1]
```diff
@@ -88,6 +88,7 @@
     <ClCompile Include="@(PyCoreSource)" />
     <ClCompile Include="@(PyExtModSource)" />
     <ClCompile Include="$(PyBaseDir)shared\readline\*.c" />
+    <ClCompile Include="$(PyBaseDir)shared\runtime\pyexec.c" />
```
Comment (stinos) [inline ports/windows/micropython.vcxproj:91]: Super minor but can you move this one line down so it's aplhabetically?

File: tests/run-tests.py [hunk 2]
```diff
@@ -232,6 +233,11 @@ def send_get(what):
                     with open(test_file, "rb") as f:
                         # instead of: output_mupy = subprocess.check_output(args, stdin=f)
                         master, slave = pty.openpty()
+                        # disable auto newline conversion on the pty terminal.
+                        attrs = termios.tcgetattr(slave)
```
Comment (stinos) [inline tests/run-tests.py:237]: You're probably aware but this doesn't exist on windows. Isn't there a way to just do a `replace("/r/n", "/n")` before comparing results, or something like that?

Comment (andrewleech) [reply tests/run-tests.py:237]: Oh, I thought `pty` itself wasn't avaialble either so it would be a moot point? 
I initially thought of just adding _another_ newline replace, but thought fixing the problem at the source was cleaner than further fiddling with test results.

File: ports/windows/micropython.vcxproj (hunk 1 above)
Comment (andrewleech) [reply ports/windows/micropython.vcxproj:91]: Ah whoops, I honestly thought I was alreadyt putting it in order, eyes blurred readline/runtime and completely missed runtime\gchelper_generic below!

File: tests/run-tests.py (hunk 2 above)
Comment (andrewleech) [reply tests/run-tests.py:237]: Yeah it looks like these repl_* tests are already skipped on windows master builds https://ci.appveyor.com/project/dpgeorge/micropython/builds/48536058/job/4o16bx3lnvsu7nbn
image

**Comment by dpgeorge** [comment]
Regarding the comment above about `raise SystemExit`: actually this PR changes the behaviour, now if you `raise SystemExit` from the REPL it does nothing.  Currently on master it will exit MicroPython back to the shell.

We need to retain the existing behaviour, so we don't break existing usage, and because that's how CPython works.

Then it's a separate question how to handle this on bare-metal targets.  But for now we shouldn't change the unix behaviour on this.

**Comment by andrewleech** [comment]
> Regarding the comment above about `raise SystemExit`: actually this PR changes the behaviour, now if you `raise SystemExit` from the REPL it does nothing. Currently on master it will exit MicroPython back to the shell.
> 
> We need to retain the existing behaviour, so we don't break existing usage, and because that's how CPython works.
> 

Thanks, yes I hadn't realised this behavior had been changed inadvertantly. I've now fixed it in the latest push.
There was an associates variable here `pyexec_system_exit` which I did a global search for - it was never actually being used (anymore) so I've stripped that out as well. 
Any variable passed to SystemExit is now returned for unix port.
```
anl@STEP: ~ $ micropython
MicroPython v1.24.0-preview.102.g03e992d59d.dirty on 2024-07-09; linux [GCC 9.4.0] version
Type "help()" for more information.
>>> raise SystemExit
anl@STEP: ~ $ echo $?
0
anl@STEP: ~ $ micropython
MicroPython v1.24.0-preview.102.g03e992d59d.dirty on 2024-07-09; linux [GCC 9.4.0] version
Type "help()" for more information.
>>> raise SystemExit(-1)
anl@STEP: ~ $ echo $?
255
```
... oh, maybe one last little bug to fix there...

> Then it's a separate question how to handle this on bare-metal targets. But for now we shouldn't change the unix behaviour on this.

Well it looks to me like this should be treated as a soft reset. 
https://github.com/micropython/micropython/blob/358e501e75c028133cff1045d2062ae31206a22c/extmod/modmachine.c#L66

The current push also works this way:
```
anl@STEP: ~/micropython $ mpremote
Connected to MicroPython at /dev/ttyACM0
Use Ctrl-] or Ctrl-x to exit this shell
MicroPython v1.24.0-preview.104.g161a9c3097.dirty on 2024-07-09; WEACTV20_F411 with STM32F411CE
Type "help()" for more information.
>>> raise SystemExit
MPY: sync filesystems
MPY: soft reboot
MicroPython v1.24.0-preview.104.g161a9c3097.dirty on 2024-07-09; WEACTV20_F411 with STM32F411CE
Type "help()" for more information.
>>>
```

**Comment by dpgeorge** [comment]
> There was an associates variable here `pyexec_system_exit` which I did a global search for - it was never actually being used (anymore) so I've stripped that out as well.

Right, it would be a nice clean up to get rid of that variable, and also change the behaviour of `SystemExit` on bare-metal ports to match unix/CPython.

Because this is a change in user-facing behaviour, I've split that change out in #15486.

**Comment by andrewleech** [comment]
> prints the parse tree out

Thanks, I wasn't aware of this development feature. I also hadn't noticed the amount of overlap between pyexec/parse_compile_execute function and unix/main/execute_from_lexer() but there's probably not much to be really gained by consolidating the two.

I've made the change now and assume it's working:
```
anl@STEP: ~/micropython $ ./ports/unix/build-coverage/micropython -v -v -vMicroPython v1.24.0-preview.187.g95bce29344.dirty on 2024-08-14; linux [GCC 9.4.0] version
Type "help()" for more information.
>>> import os
----------------
[   1] import_name(14) (n=1)
         id(os)
----------------
File <stdin>, code block '<module>' (descriptor: 7f9fea2a9320, bytecode @7f9fea2a9180 11 bytes)
Raw bytecode (code_info_size=3, bytecode_size=8):
 08 02 01 80 51 1b 02 16 02 51 63
arg names:
(N_STATE 2)
(N_EXC_STACK 0)
  bc=0 line=1
00 LOAD_CONST_SMALL_INT 0
...
```

**Comment by dpgeorge** [comment]
It looks like there are still a couple of things missing in pyexec.c that the unix port has:
- support the `compile_only` flag
- set `__file__` when `MICROPY_PY___FILE__` is enabled

**Comment by dpgeorge** [comment]
> It looks like there are still a couple of things missing in pyexec.c that the unix port has

This is why #6008 is quite a bit bigger than this PR, and why it was still in progress, because of all these details :smile:

**Comment by dpgeorge** [comment]
@andrewleech there are still two of my outstanding points above to address, and also @stinos comment about newlines.  Did you want to address those?  Alternatively I can merge #16141 now to get raw REPL support on the unix port, then this PR can go in after that, when it's ready.

**Comment by andrewleech** [comment]
I would very much like to get this PR finished to unblock aiorepl use, I'll reprioritize it in my backlog cleanups effort! 

Previously I didn't like the idea of a blanket replace it newlines like @stinos suggested however after battling inconsistent newline matching in some cmdline tests just the other day I'm now far more interested in the simple/broad fix!

**Comment by andrewleech** [comment]
@dpgeorge this should be ready for re-review.

This PR does not include the unhandled exception hooks that you had in https://github.com/micropython/micropython/pull/6008 but I feel that should be a separate feature / pr anyway if it's still needed by others?

There's a minor compatibility update to aiorepl in https://github.com/micropython/micropython-lib/pull/1016 which has been tested with this PR and seems to behave as expected. Without that change aiorepl needs crtl-enter to execute / send a line.

File: shared/runtime/pyexec.c [hunk 3]
```diff
@@ -116,7 +131,9 @@ static int parse_compile_execute(const void *source, mp_parse_input_kind_t input
         #if MICROPY_REPL_INFO
         start = mp_hal_ticks_ms();
         #endif
-        mp_call_function_0(module_fun);
+        if (!(exec_flags & EXEC_FLAG_COMPILE_ONLY)) {
```
Comment (dpgeorge) [inline shared/runtime/pyexec.c]: This compile-only option doesn't work correctly at the REPL, eg:
```
$ ./build-standard/micropython -X compile-only
MicroPython v1.26.0-preview.282.gdcbdeb839 on 2025-06-27; linux [GCC 15.1.1] version
Type "help()" for more information.
>>> 1/2
0.5
>>> 
```
That's wrong, it shouldn't print anything as the answer.

I suggest to just keep this simple and add a compile-time option to enable this:
```c
#if MICROPY_PYEXEC_COMPILE_ONLY
if (mp_compile_only)
#endif
{
    mp_call_function_0(module_fun);
}
```
Then the unix port enables this and makes `mp_compile_only` globally available and declared in `mphalport.h`.

That will mean `pyexec_file_with_flags()` isn't needed, and ports that don't use this feature won't take a hit.

Comment (andrewleech) [reply shared/runtime/pyexec.c]: Thanks for the suggestion, I've implemented this and pushed.

**Comment by dpgeorge** [comment]
@andrewleech can you please check the CI failures?  Not sure what's going on there.

**Comment by dpgeorge** [comment]
Note: the code size increase across most of the ports is due to improved support for `MICROPY_PY___FILE__`: the variable `__file__` is now set whenever a script is executed via pyexec.  So now we have:
- `__file__` is already set on import of scripts (eg `import foo`, then `__file__` would be `foo.py`), that doesn't change with this PR
- when executing boot.py/main.py from the filesystem, `__file__` is set to either `"boot.py"` or `"main.py"` respectively (this is not done for frozen boot.py/main.py though)
- doing `mpremote run test.py` now sets `__file__` to `"<stdin>"`
- at the REPL, `__file__` now exists and is the name of the most recently executed file

I think that's a good improvement, makes things more consistent.

File: shared/runtime/pyexec.c [hunk 4]
```diff
@@ -64,7 +65,7 @@ static bool repl_display_debugging_info = 0;
 // EXEC_FLAG_ALLOW_DEBUGGING allows debugging info to be printed after executing the code
 // EXEC_FLAG_IS_REPL is used for REPL inputs (flag passed on to mp_compile)
 static int parse_compile_execute(const void *source, mp_parse_input_kind_t input_kind, mp_uint_t exec_flags) {
-    int ret = 0;
+    mp_int_t ret = 0;
```
Comment (dpgeorge) [inline shared/runtime/pyexec.c]: I think this `ret` variable should stay as `int`, since that's the return type of this function.

File: ports/unix/Makefile [hunk 5]
```diff
@@ -42,6 +42,7 @@ GIT_SUBMODULES += lib/berkeley-db-1.xx
 INC +=  -I.
 INC +=  -I$(TOP)
 INC += -I$(BUILD)
+INC += -I$(TOP)/shared/runtime
```
Comment (dpgeorge) [inline ports/unix/Makefile]: This include shouldn't be needed.

File: ports/unix/main.c [hunk 6]
```diff
@@ -297,18 +234,25 @@ static int do_repl(void) {
             line = line3;
         }
 
-        int ret = execute_from_lexer(LEX_SRC_STR, line, MP_PARSE_SINGLE_INPUT, true);
+        ret = execute_from_lexer(LEX_SRC_STR, line, MP_PARSE_SINGLE_INPUT, true);
         free(line);
-        if (ret & FORCED_EXIT) {
-            return ret;
-        }
```
Comment (dpgeorge) [inline ports/unix/main.c:308]: I think is `if` needs to stay, otherwise `sys.exit()` doesn't work.

File: ports/unix/mphalport.h [hunk 7]
```diff
@@ -48,6 +49,12 @@ void mp_hal_set_interrupt_char(char c);
 void mp_hal_stdio_mode_raw(void);
 void mp_hal_stdio_mode_orig(void);
 
+// pyexec/repl needs stdio to be in raw mode, but this may be cleared before running code.
+#if MICROPY_REPL_RESET_RAW_BEFORE_EXEC
```
Comment (dpgeorge) [inline ports/unix/mphalport.h]: I don't think this is used anymore??

File: ports/unix/main.c [hunk 8]
```diff
@@ -297,18 +234,25 @@ static int do_repl(void) {
...
+
 static int do_file(const char *file) {
-    return execute_from_lexer(LEX_SRC_FILENAME, file, MP_PARSE_FILE_INPUT, false);
+    int ret = pyexec_file(file);
+    // pyexec returns 1 for success, 0 for exception, PYEXEC_FORCED_EXIT for SystemExit
+    // Convert to unix port's expected codes: 0 for success, 1 for exception, FORCED_EXIT|val for SystemExit
+    if (ret == 1) {
+        return 0; // success
+    } else if (ret == PYEXEC_FORCED_EXIT) {
+        return FORCED_EXIT; // SystemExit (without exit value for now)
```
Comment (dpgeorge) [inline ports/unix/main.c]: The value of `SystemExit` must be propagated out of this function, otherwise `sys.exit(1)` doesn't return an error code to the shell.

**Comment by andrewleech** [comment]
@dpgeorge I've addressed your comments and added a couple of basic tests for systemexit - the outputs aren't very interesting but basically show a number in systemexit is shown as a failed (crash) test. Zero or no arg is a successful result.

File: shared/runtime/pyexec.c [hunk 9]
```diff
@@ -141,8 +160,13 @@ static int parse_compile_execute(const void *source, mp_parse_input_kind_t input
 
         // check for SystemExit
         if (mp_obj_is_subclass_fast(MP_OBJ_FROM_PTR(((mp_obj_base_t *)nlr.ret_val)->type), MP_OBJ_FROM_PTR(&mp_type_SystemExit))) {
-            // at the moment, the value of SystemExit is unused
-            ret = PYEXEC_FORCED_EXIT;
+            // Extract SystemExit value
```
Comment (dpgeorge) [inline shared/runtime/pyexec.c]: It'd be worth copying the comment over from the unix port's implementation of this:
```
// None is an exit value of 0; an int is its value; anything else is 1
```

File: ports/unix/main.c [hunk 10]
```diff
@@ -297,18 +234,28 @@ static int do_repl(void) {
...
+    if (ret == 1) {
+        return 0; // success
+    } else if (ret & PYEXEC_FORCED_EXIT) {
+        return ret; // SystemExit with exit value in lower 8 bits
+    } else {
+        return 1; // exception
+    }
 }
 
 static int do_str(const char *str) {
```
Comment (dpgeorge) [inline ports/unix/main.c:272]: It would be possible to replace the `execute_from_lexer()` below with `pyexec_vstr()` (just make a temporary vstr on the C stack and populate it with `str` and `strlen(str)`).

Might be worth doing that as part of this PR?

**Comment by Josverl** [comment]
Is there a specific reason for
> this is not done for frozen boot.py/main.py though

**Comment by andrewleech** [comment]
> Is there a specific reason for
> > this is not done for frozen boot.py/main.py though

I'm assuming the `__FILE__` setting section is in a different pathway to the "run from frozen" code path. This wasn't intentional certainly. It might not be that easy to fix in a generic way though as I presume frozen code will be referenced by python path rather than the typical file path that's used here.

**Comment by dpgeorge** [comment]
> I'm assuming the `__FILE__` setting section is in a different pathway to the "run from frozen" code path. This wasn't intentional certainly. It might not be that easy to fix in a generic way though as I presume frozen code will be referenced by python path rather than the typical file path that's used here.

Yes, that's correct.  There would need to be extra code added to set `__FILE__` for execution of frozen boot/main.

**Comment by andrewleech** [comment]
> Yes, that's correct. There would need to be extra code added to set __FILE__ for execution of frozen boot/main.

I'm thinking that would be worth doing too for consistency at runtime, would it make sense to just report these frozen modules as `<module name>.py` ? 
Or would it be better as `frozen:<module name>.py` or similar so users can tell if a frozen version is being used?

**Comment by dpgeorge** [comment]
> I'm thinking that would be worth doing too for consistency at runtime, would it make sense to just report these frozen modules as `<module name>.py` ?

There's code in `py/builtinimport.c` that does it already for import of frozen modules.  Following that logic would be best.

**Comment by Josverl** [comment]
as a side note; if all frozen modules can indeed be identified using the pattern `<module name>.py`  that would also allow them to be excluded by default from sys.settrace tracing,

File: ports/unix/main.c (hunk 10 above)
Comment (andrewleech) [reply ports/unix/main.c:272]: `execute_from_lexer` is also still used for a couple of other things. I've trimmed it down a bit to remove now-unused functionality but would it be better refactored into a couple of new `pyexec_str_single()` and `pyexec_stdin()` functions in pyexec just to consolidate the location of all this sort of code? These functions are only needed on desktop really so would probably want extra #defines around them.

Comment (dpgeorge) [reply ports/unix/main.c:272]: > but would it be better refactored into a couple of new `pyexec_str_single()` and `pyexec_stdin()` functions in pyexec just to consolidate the location of all this sort of code?

Yes, I think we should do that and get rid of `execute_from_lexer`.  I was going to do that in a follow-up PR, but if you want to do it here, feel free.



> These functions are only needed on desktop really so would probably want extra #defines around them.

Yes.

File: ports/unix/main.c [hunk 11]
```diff
@@ -297,22 +226,43 @@ static int do_repl(void) {
...
-    return execute_from_lexer(LEX_SRC_STR, str, MP_PARSE_FILE_INPUT, false);
+    vstr_t vstr;
+    vstr_init(&vstr, strlen(str));
+    vstr_add_strn(&vstr, str, strlen(str));
```
Comment (dpgeorge) [inline ports/unix/main.c]: This allocates RAM unnecessarily.  I suggest using `vstr_init_fixed_buf()`.

File: ports/unix/main.c [hunk 12]
```diff
@@ -297,22 +226,43 @@ static int do_repl(void) {
...
+    int ret = pyexec_vstr(&vstr, false);
+    vstr_clear(&vstr);
+    if (ret == 1) {
+        return 0; // success
+    } else if (ret & PYEXEC_FORCED_EXIT) {
+        return ret; // SystemExit with exit value in lower 8 bits
+    } else {
+        return 1; // exception
+    }
```
Comment (dpgeorge) [inline ports/unix/main.c]: This code should be factored/shared with `do_file()`.

Comment (andrewleech) [reply ports/unix/main.c]: Updated as suggested thanks

File: ports/unix/main.c (hunk 11 above)
Comment (andrewleech) [reply ports/unix/main.c]: Updated as suggested thanks

File: ports/unix/main.c (hunk 10 above)
Comment (andrewleech) [reply ports/unix/main.c:272]: I've done this as well now thanks

**Comment by dpgeorge** [comment]
@andrewleech can you please rebase this on latest master?  In particular it will bring in the test from a791f7f43ccc21f4275e3905240e61cf5bfa1c0c which is important here.

**Comment by andrewleech** [comment]
> @andrewleech can you please rebase this on latest master

done

**Comment by andrewleech** [comment]
@dpgeorge the branch has been rebased and newer tests updated

File: shared/runtime/pyexec.c [hunk 13]
```diff
@@ -85,6 +86,14 @@ static int parse_compile_execute(const void *source, mp_parse_input_kind_t input
             ctx->module.globals = mp_globals_get();
             ctx->constants = frozen->constants;
             module_fun = mp_make_function_from_proto_fun(frozen->proto_fun, ctx, NULL);
+
+            #if MICROPY_MODULE___FILE__
+            // Set __file__ for frozen MPY modules
+            if (input_kind == MP_PARSE_FILE_INPUT && frozen_module_name != NULL) {
```
Comment (andrewleech) [inline shared/runtime/pyexec.c]: Check if this can ever be false?

File: ports/unix/main.c (hunk 11 above)
Comment (dpgeorge) [reply ports/unix/main.c]: I think you forgot to commit this change?

File: ports/unix/main.c (hunk 12 above)
Comment (dpgeorge) [reply ports/unix/main.c]: I think you forgot to commit this change?

**Comment by dpgeorge** [comment]
I found that ctrl-C didn't work on the unix port :grimacing: 

I fixed that and pushed the commit to this branch.  @andrewleech see what you think, it was a little tricky to get right.

**Comment by dpgeorge** [comment]
It looks like my fix for ctrl-C handling has slightly changed the output for using paste mode, probably due to switch in/out of raw mode.

@andrewleech do you think we can just change the tests .exp file to match the new behaviour?  Or try to make it go back to the previous behaviour?

**Comment by andrewleech** [comment]
Thanks @dpgeorge  I looked into the mode switching and agree it's required, I couldn't find any other way to make paste mode and keyboard interrupt both work. I've consolidated that in a function with comments, see if you think that helps?

 I also added some new tests, one of which is sitting in the unix tests folder but needs to be run directly with cpython as a standalone script  to test interactive ctrl-c handling on the unix port. It failed without your mode switch patch and passes now, it's a bit messy but validates the behaviour. It's not automatically run as part of the test suite though. Does this look useful and should it be integrated better?

I've found and included the other consolidation of do_file and do_str results handling. 

I thought I had the unit test results cleaned up too but it appears not

File: tests/ports/unix/test_ctrl_c_interrupt.py [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+                break
+            output += data
+        else:
+            break
+    return output
+
+
+def main():
+    # Get micropython path from command line or use default
+    micropython = sys.argv[1] if len(sys.argv) > 1 else "../ports/unix/build-standard/micropython"
```
Comment (dpgeorge) [inline tests/ports/unix/test_ctrl_c_interrupt.py]: This executable will somehow need to be customised, eg for minimal, coverage, nanbox builds/tests.

File: shared/runtime/pyexec.c [hunk 15]
```diff
@@ -565,10 +590,11 @@ int pyexec_raw_repl(void) {
             // exit for a soft reset
             mp_hal_stdout_tx_str("\r\n");
             vstr_clear(&line);
+            mp_hal_stdio_mode_orig();
             return PYEXEC_FORCED_EXIT;
         }
 
-        int ret = parse_compile_execute(&line, MP_PARSE_FILE_INPUT, EXEC_FLAG_PRINT_EOF | EXEC_FLAG_SOURCE_IS_VSTR);
+        int ret = exec_with_kbd_intr_support(&line, MP_PARSE_FILE_INPUT, EXEC_FLAG_PRINT_EOF | EXEC_FLAG_SOURCE_IS_VSTR);
```
Comment (dpgeorge) [inline shared/runtime/pyexec.c]: Suggest not having this helper function, and just writing out the code here directly.  Otherwise it's too hard to follow the logic for `(ret & PYEXEC_FORCED_EXIT)`.

File: tests/ports/unix/test_ctrl_c_interrupt.py (hunk 14 above)
Comment (andrewleech) [reply tests/ports/unix/test_ctrl_c_interrupt.py]: I rebuild this test to integrate in the test runner properly with some modifications to the repl_ test handler in 66c748ca24c9d93e0b6e973bc4b4187d354d6d18

**Comment by andrewleech** [comment]
The unix ctrl-c tests have been split out from this PR again into https://github.com/micropython/micropython/pull/18435

**Review by dpgeorge — APPROVED** [review]
I tested this thoroughly and fixed a minor issue: ctrl-C was not properly handled when executing a string via `micropython -c "import time; time.sleep(10)"`.

**Comment by andrewleech** [comment]
Thanks for your assistance and detailed testing!
